### PR TITLE
plugins/null-ls: refactoring + tests

### PR DESCRIPTION
**Pull request by GaetanLepage**
Refactoring of null-ls: added more options.
I also wrote tests.
Fixes https://github.com/pta2002/nixvim/issues/32.

### CODE REVIEW

File: plugins/null-ls/default.nix [hunk 1]
```diff
@@ -12,38 +12,244 @@ in {
...
+
+        Note that by default, Neovim will not display updated diagnostics in insert mode.
+        Together with the option above, you need to pass `update_in_insert = true` to
+        `vim.diagnostic.config` for diagnostics to work as expected.
+        See `:help vim.diagnostic.config` for more info.
+      '';
+
+      sourcesItems = mkOption {
+        default = null;
+        # type = with types; nullOr (either (listOf str) (listOf attrsOf str));
```
Comment (traxys) [inline plugins/null-ls/default.nix:202]: I think this is a mistake

File: plugins/null-ls/default.nix [hunk 2]
```diff
@@ -12,38 +12,244 @@ in {
...
+        type = with types; nullOr (listOf (attrsOf str));
+        description = "The list of sources to enable, should be strings of lua code. Don't use this directly";
+      };
+
+      # sources = mkOption {
+      #   default = null;
+      #   type = with types; nullOr attrs;
+      # };
```
Comment (traxys) [inline plugins/null-ls/default.nix:210]: Also this one

File: plugins/null-ls/default.nix [hunk 3]
```diff
@@ -12,38 +12,244 @@ in {
...
+          (helpers.mkRaw cfg.onAttach);
+        on_init =
+          helpers.ifNonNull' cfg.onInit
+          (helpers.mkRaw cfg.onInit);
+        on_exit =
+          helpers.ifNonNull' cfg.onExit
+          (helpers.mkRaw cfg.onExit);
+        root_dir =
+          helpers.ifNonNull' cfg.rootDir
+          (helpers.mkRaw cfg.rootDir);
```
Comment (traxys) [inline plugins/null-ls/default.nix:243]: We should  maybe look into a helper for this in the future

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/null-ls/default.nix (hunk 1 above)
Comment (GaetanLepage) [reply plugins/null-ls/default.nix]: Fixed !

File: plugins/null-ls/default.nix (hunk 3 above)
Comment (GaetanLepage) [reply plugins/null-ls/default.nix:236]: yes it could be useful...

**Review by traxys — APPROVED** [review]

**Pull request merged**